addons/pinniped/post-deploy/pkg/configure/configure.go

```diff
@@ -178,11 +179,83 @@ func TKGAuthentication(c Clients) error {
+	tkgMetadata := &inspect.TKGMetadata{}
+	tkgMetadata.Cluster = struct {
+		Name           string `yaml:"name"`
+		Type           string `yaml:"type"`
+		Infrastructure struct {
+			Provider string `yaml:"provider"`
+		} `yaml:"infrastructure"`
+	}{}
+
+	// TKGs systems will:
+	// - pass --jwtauthenticator-audience=<cluster.name>-<cluster.id>
+	// - therefore we can infer a "workload" cluster in this case
+	// TKGm systems will currently:
+	// - provide tkg-metadata configmap for cluster.name and cluster.type
+	// TKGm sytems in the future:
+	// - will not provide the tkg-metadata configmap.
+	// - therefore
+	//    - we can use the node.Annotations["cluster.x-k8s.io/cluster-name"] for cluster.name
+	//    - we do not know yet how to find the cluster.type
+	//      TODO: we can likely use
+	//      - https://github.com/vmware-tanzu/tanzu-framework/blob/main/docs/api-machinery/capability-discovery.md
+	//      - https://github.com/vmware-tanzu/tanzu-framework/blob/main/pkg/v1/sdk/capabilities/discovery/tkg/capabilities.go#L29
+	//      to find more details about cluster.type, the c.IsWorkloadCluster() and c.IsManagementCluster() and c.IsTKGm()
+	//      may be solutions to this problem.
+	if len(vars.JWTAuthenticatorAudience) == 0 {
+		// get the tkgMetadata configmap. if we can't, then we need to fake it out as best we can.
+		foundMetadata, err := inspector.GetTKGMetadata()
+		if err != nil {
+			// lets log it, but try getting what we need from an annotation on the node
+			zap.S().Error(err)
+			nodeList, listError := c.K8SClientset.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
+			if listError != nil {
+				zap.S().Error(listError)
+				// if we can't list nodes, we don't have an alternative course of action at this point.
+				// fail hard. this is an unexpected fail state. we could look into alternatives such as:
+				// - get the kube-system namespace for the cluster and use its UID
+				// - simply generate another random string
+				return listError
+			}
+			if len(nodeList.Items) == 0 {
+				zap.S().Error("no nodes available to read cluster-name annotation")
+				return stdliberrors.New("no nodes available to read cluster-name annotation")
+			} else {
+				zap.S().Info("Successfully listed Nodes")
+				// cluster-api provides cluster-name annotation on nodes
+				var clusterName = ""
+				for _, node := range nodeList.Items {
+					if len(node.Annotations["cluster.x-k8s.io/cluster-name"]) != 0 {
+						clusterName = node.Annotations["cluster.x-k8s.io/cluster-name"]
+						break
+					}
+				}
+				if clusterName == "" {
+					return stdliberrors.New("nodes not annotated with cluster-name annotation")
+				} else {
+					tkgMetadata.Cluster.Name = clusterName
+					// TODO: we still need the matching tkgMetadata.Cluster.Type here
+					// once tkg-metadata is no longer available. The post-deploy job will
+					// fail and throw an error once the cm disappears.
+				}
+			}
+		} else {
+			tkgMetadata = foundMetadata
```

Comment:
Think I might be missing something since I'm not super familiar with this code base, but in the case where you're looking up tkg-metadata or the K8s node annotations...how is audience passed back to the call to Pinniped() below? I see that's always using 'vars.JWTAuthenticatorAudience'

Reply:
within the `Pinniped()` there is a block that does:

```goland
audience := p.JWTAuthenticatorAudience
if audience == "" {
	audience = p.ClusterName
}
```
its a little messy but seems to get the job done.